### PR TITLE
[Merged by Bors] - support non routable nodes

### LINKED ISSUE #1697: Support non-routable nodes
Add a fallback mechanism to support non-routable nodes. These nodes will tcp connect to other nodes and ask neighbors. They should be provided with neighbors so they can connect and sync but they should not be given as neighbors to other nodes as these other nodes can't connect to them.

### PR DESCRIPTION
**Pull request by y0sher**
## Motivation
Closes #1697

## Changes

uses src port tuple to return udp messages, uses `connect` syscall for udp abstraction

## Test Plan
added unit tests
## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

### CODE REVIEW

**Comment by y0sher** [comment]
bors try

File: p2p/integration_test.go [hunk 1]
```diff
@@ -133,7 +133,6 @@ func Test_SmallP2PIntegrationSuite(t *testing.T) {
 }
 
 func Test_BigP2PIntegrationSuite(t *testing.T) {
-	t.Skip()
```
Comment (noamnelke) [inline p2p/integration_test.go]: let's add:
```golang
if testing.Short {
	t.Skip()
}
```

File: p2p/net/msgcon.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+// Closed returns whether the connection is closed
+func (c *MsgConnection) Closed() bool {
+	c.wmtx.Lock()
+	defer c.wmtx.Unlock()
+	return c.closed
+}
+
+//func (c *MsgConnection) setupIncoming(timeout time.Duration) error {
```
Comment (noamnelke) [inline p2p/net/msgcon.go]: also left here intentionally? if yes, add a comment?

File: p2p/net/network_mock.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ func (m ReadWriteCloserMock) Close() error {
 	return nil
 }
 
-//RemoteAddr mocks remote addr return
+//RemoteAddr mocks remote Addr return
```
Comment (noamnelke) [inline p2p/net/network_mock.go]: over refactoring? not critical, though

File: p2p/net/udp.go [hunk 4]
```diff
@@ -60,7 +106,7 @@ func (n *UDPNet) Start() error {
 	return nil
 }
 
-// LocalAddr returns the local listening addr, will panic before running Start. or if start errored
+// LocalAddr returns the local listening Addr, will panic before running Start. or if start errored
```
Comment (noamnelke) [inline p2p/net/udp.go]: same?

File: p2p/net/udp.go [hunk 5]
```diff
@@ -109,10 +155,44 @@ func (n *UDPNet) Send(to *node.NodeInfo, data []byte) error {
...
 
 // IncomingMessages is a channel where incoming UDPMessagesEvents will stream
-func (n *UDPNet) IncomingMessages() chan UDPMessageEvent {
+func (n *UDPNet) IncomingMessages() chan IncomingMessageEvent {
 	return n.msgChan
 }
 
+//Dial(address inet.Addr, remotePublicKey p2pcrypto.PublicKey) (net.Connection, error) // Connect to a remote node. Can send when no error.
+//SubscribeOnNewRemoteConnections(func(event net.NewConnectionEvent))
+//SubscribeClosingConnections(func(net.ConnectionWithErr))
```
Comment (noamnelke) [inline p2p/net/udp.go]: remove?

File: p2p/net/udp.go [hunk 6]
```diff
@@ -140,33 +220,163 @@ func (n *UDPNet) listenToUDPNetworkMessages(listener net.PacketConn) {
...
+
+			if err != nil {
+				n.logger.Warning("error can't extract public key from udp message. (Addr=%v), err=%v", addr.String(), err)
+				continue
+			}
+
+			ns := n.cache.GetOrCreate(pk)
+
+			if ns == nil {
+				n.logger.Warning("coul'd not create session with %v:%v skipping message..", addr.String(), pk.String())
```
Comment (noamnelke) [inline p2p/net/udp.go]:
```suggestion
				n.logger.Warning("could not create session with %v:%v skipping message..", addr.String(), pk.String())
```

File: p2p/net/udp.go [hunk 7]
```diff
@@ -140,33 +220,163 @@ func (n *UDPNet) listenToUDPNetworkMessages(listener net.PacketConn) {
...
+				continue
+			}
+
+			host, port, err := net.SplitHostPort(addr.String())
+			if err != nil {
+				n.logger.Warning("coudl'nt parse address skipping message from  %v %V", addr.String(), pk)
```
Comment (noamnelke) [inline p2p/net/udp.go]:
```suggestion
				n.logger.Warning("couldn't parse address skipping message from  %v %V", addr.String(), pk)
```

Comment (noamnelke) [reply p2p/net/udp.go]: also: consider using `%x` for the public key (assuming it's a byte-array)

File: p2p/net/msgcon.go [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	deadliner   deadliner
+	messages    chan []byte
+	stopSending chan struct{}
+	close       io.Closer
+
+	msgSizeLimit int
+}
+
+// Create a new connection wrapping a net.Conn with a provided connection manager
+func newMsgonnection(conn readWriteCloseAddresser, netw networker,
```
Comment (noamnelke) [inline p2p/net/msgcon.go]: `newMsgonnection` -> `newMsgConnection`?

File: p2p/udp.go [hunk 9]
```diff
@@ -22,9 +23,11 @@ type Lookuper func(key p2pcrypto.PublicKey) (*node.NodeInfo, error)
...
 	Start() error
 	Shutdown()
-
-	IncomingMessages() chan inet.UDPMessageEvent
-	Send(to *node.NodeInfo, data []byte) error
+	Dial(address net.Addr, remotePublicKey p2pcrypto.PublicKey) (inet.Connection, error)
+	IncomingMessages() chan inet.IncomingMessageEvent
+	SubscribeOnNewRemoteConnections(f func(event inet.NewConnectionEvent))
+	SubscribeClosingConnections(f func(connection inet.ConnectionWithErr))
+	//Send(to *node.NodeInfo, data []byte) error
```
Comment (noamnelke) [inline p2p/udp.go]: delete?

File: p2p/udp.go [hunk 10]
```diff
@@ -191,33 +225,57 @@ func (upm *udpProtocolMessage) Data() service.Data {
...
+	if err != nil {
+		mux.logger.Warning("failed decrypting message err=%v", err)
+		return ErrFailDecrypt
+	}
+
+	pm := &ProtocolMessage{}
+	err = types.BytesToInterface(decPayload, pm)
 	if err != nil {
-		return errors.New("could'nt deserialize message")
+		mux.logger.Error("proto marshaling err=", err)
```
Comment (noamnelke) [inline p2p/udp.go]:
```suggestion
		mux.logger.Error("deserialization err=", err)
```
(this is XDR deserialization from bytes into a struct defined using protobuf, or am I missing something?)

**Review by noamnelke — APPROVED** [review]
I only had very minor comments 👍

File: p2p/net/udp.go (hunk 7 above)
Comment (y0sher) [reply p2p/net/udp.go]: it's a bytearay with a `String` so there's no need.

File: p2p/udp.go (hunk 10 above)
Comment (y0sher) [reply p2p/udp.go]: it was just leftovers from when we used protobufs

**Comment by y0sher** [comment]
I addressed all your comments.

**Comment by y0sher** [comment]
bors merge

**Comment by y0sher** [comment]
bors merge

**Pull request closed without merging**